### ISSUE TITLE
Problems with interrupt

### ISSUE THREAD

**Issue by nickehallgren** [eta-systems/MAX7313-Arduino#3]
Hi and a big thanks for this library. It works very welll but I havn't got the interrupt to work. I have the MAX7313 INT pin (1) connected to the Raspberry Pi Pico (GPIO 2) with a pullup to 3.3V (as I understand the INT should go low when a change is noticed).

```
pinMode(2, INPUT_PULLUP);
attachInterrupt(2, trigged, FALLING );

void trigged(){
  Serial.println("Change happend!");
}
```
the max code (to enable interrupt)
```
pwm.pinMode(8, INPUT);
pwm.pinMode(16, OUTPUT);  // Tried without this too
pwm.enableInterrupt();
```
but nothing happens when I press a button on port 8, if I read the port (digitalRead(8)) I can see when the button is pressed. What am I missing?

**Comment by mnemocron** [comment]
Hey,
I haven't worked with the chip in a while and I am not sure if I tested the interrupt functionality.
I just looked at the library code and am confident, that the interrupt config is correct.
Unfortunately, I currently don't have a MAX7313 around to debug this. 

However, you could try to use **`clearInterrupt()`**

```cpp
pwm.enableInterrupt();
pwm.clearInterrupt();  // to acknowledge that you received the interrupt

void trigged(){
  pwm.clearInterrupt();  // must be called each time
  Serial.println("Change happend!");
}
```

(i should add this as an example to the repo)

If that still does not change the behaviour, tell me and I might ask you to check the INT/O16 signal with a probe.
In the meantime I will try to get my hands on a MAX7313.

**Comment by mnemocron** [comment]
Alright, I found some example code (not the Arduino lib)

pwm.[enableInterrupt()](https://github.com/eta-systems/MAX7313/blob/master/example/Src/main.c#L150) after `pwm.begin()` (aka. `init()`)

and then clearing the interrupt every time it occurs inside the ISR:
pwm.[clearInterrupt()](https://github.com/eta-systems/MAX7313/blob/master/example/Src/stm32f3xx_it.c#L197)

So I think, this worked at some point.

**Comment by nickehallgren** [comment]
Thanks for the fast resonse. I actually already tried with clearInterrupt() at setup and in trigged but without success...

**Comment by nickehallgren** [comment]
Well this is wierd, I pulled up the scope and connected it to INT and then I saw the pulse, and now it just works... Must have been a loose wire on the breadboard even if I'm sure I already checked that... But yes it works now, thanks!

**Comment by nickehallgren** [comment]
Ok, so not so simple. I get it to work sometimes after I disconnect/reconnect the INT while running, but when the interrupt starts to work I get no pwm output anymore (have a RGB led connected to 9-11).

here's my sketch as of now, 
```
#include <Wire.h>
#include "max7313.h"

const byte PICO_I2C_SDA = 0;
const byte PICO_I2C_SCL = 1;

MAX7313 brd1(0x20);  // on I2C address 0x20

void setup() {
  Serial.begin(115200);
  // Pico setup
  pinMode(2, INPUT_PULLUP);
  attachInterrupt(2, trigged, FALLING );
  Wire.setSDA(PICO_I2C_SDA);
  Wire.setSCL(PICO_I2C_SCL);

  brd1.begin();
  brd1.enableInterrupt();
  brd1.clearInterrupt();
  brd1.pinMode(8, INPUT);
  brd1.pinMode(9, OUTPUT);
  brd1.pinMode(10, OUTPUT);
  brd1.pinMode(11, OUTPUT);
}

void trigged(){
  brd1.clearInterrupt();
  Serial.println("Change happend!");
}

void loop(){
    brd1.analogWrite(9, 12);            // R
    brd1.analogWrite(10, 7);            // G
    brd1.analogWrite(11, 2);            // B
}
```

**Comment by mnemocron** [comment]
What do you mean by no PWM output anymore? The signal is low/high without PWM?
Maybe add a `delay(100)` in the loop. Otherwise, the I2C bus is constantly busy.
Maybe also consider an external pull-up for the INT signal.

**Comment by nickehallgren** [comment]
When (I just managed to get it to work a few times) I get the INT to work the RGB LED (set to white now) turns off even if its in the loop (had a delay earlier but added it back now) and I already have an external 4K7 to 3.3V on INT. Added this to the loop so I can monitor that the MAX7313 really sees the button press

```
uint8_t val = brd1.digitalRead(8);
if (lastState != val){
  Serial.print("Button state: ");
  Serial.println(val);
  lastState = val;
}
delay(100);
```

and the serial monitor confirms that it reads the button press

**Comment by mnemocron** [comment]
Just checked the datasheet again. It should work with the right configuration.
Although unlikely, a "reset" behaviour of the LED could also be a short circuit whenever the INT is triggered. This would reset the GPIOs to inputs and without the `begin()` nothing is changed. 
How does your INT signal behave if it does work? Does it go high again? Polling with digitalRead should always clear the INT flag.

Maybe I can get my hands on a MAX7313 next week. Then I can try to reproduce it.

**Comment by nickehallgren** [comment]
So after playing with the scope for a while I can see that the INT pin stays low when `enableInterrupt()` isn't called and stays high when it is as it should. But nothing happens to the INT pin when I press the button (the serial monitor confirms that the button was pressed). 

When I get it to work (reset?) I clearly see the spike to low and then it goes high again and it works until I reset the board (but with no outputs). 

And yes everything seems to be according to the datasheet, I'll keep trying and hopefully you get your hands on a MAX7313, thanks!

**Comment by mnemocron** [comment]
Hello, 
I just assembled a board and programmed it using an Arduino Due and the Arduino IDE.
I cannot reproduce the issue. PWM and interrupt work both at the same time.

Check out the code I used: [examples/Interrupt.ino](https://github.com/eta-systems/MAX7313-Arduino/blob/master/examples/Interrupt/Interrupt.ino)
But I may have fiddled around in the MAX7313.cpp configuration of the chip. 
I would clone/pull the latest version 1.0.2 from the master branch to be sure you have the same library version as I do.
And I hope this also works in your application.

**Comment by nickehallgren** [comment]
Already, wow! I tried with the latest version and your example (changed input/output ports) but the result is the same. 

Then I dug up an old UNO and tried that but now I'm confused. Nothing happens when using this line `attachInterrupt(digitalPinToInterrupt(IRQ_PIN), GPIO_ISR, FALLING);` its like everything stops, if I uncomment the line I see PWM looping. I also tried replacing the `digitalPinToInterrupt(IRQ_PIN)` with only IRQ_PIN (and with 0) and then the PWM works again but with no interrupt. I also tried printing `digitalPinToInterrupt(IRQ_PIN)` and it shows interrupt 0 as it should...
